plotly/tests/test_optional/test_figure_factory.py

```diff
@@ -2184,3 +2183,503 @@ def test_valid_facet_grid_fig(self):
         # layout
         self.assert_dict_equal(test_facet_grid['layout'],
                                exp_facet_grid['layout'])
+
+
+class TestBullet(NumpyTestUtilsMixin, TestCase):
+
+    def test_df_as_list(self):
+        df = [
+            {'title': 'Revenue'},
+            'foo'
+        ]
+
+        pattern = ('If your data is a list, all entries must be dictionaries.')
+        self.assertRaisesRegexp(PlotlyError, pattern, ff.create_bullet, df)
+
+    def test_not_df_or_list(self):
+        df = 'foo'
+
+        pattern = ('You must input a pandas DataFrame or a list of dictionaries.')
+        self.assertRaisesRegexp(PlotlyError, pattern, ff.create_bullet, df)
+
+    def test_valid_keys(self):
+        df = [{'title': 'Revenue', 'foo': 'bar'}]
+        VALID_KEYS = ['title', 'subtitle', 'ranges', 'measures', 'markers']
+
+        pattern = (
+            'Your headers/dict keys must be either {}'
+        ).format(utils.list_of_options(VALID_KEYS, 'or'))
+        self.assertRaisesRegexp(PlotlyError, pattern, ff.create_bullet, df)
+
+    def test_valid_color_lists_of_2_rgb_colors(self):
+        df = [
+            {'title': 'Revenue'}
+        ]
+
+        range_colors = ['rgb(0, 0, 0)']
+        measure_colors = ['rgb(0, 0, 0)']
+
+        pattern = ("Both 'range_colors' or 'measure_colors' must be a list "
+                   "of two valid colors.")
+        self.assertRaisesRegexp(
+            PlotlyError, pattern, ff.create_bullet, df,
+            range_colors=range_colors
+        )
+
+        self.assertRaisesRegexp(
+            PlotlyError, pattern, ff.create_bullet, df,
+            measure_colors=measure_colors
+        )
+
+    def test_full_bullet(self):
+        df = pd.read_json('https://cdn.rawgit.com/plotly/datasets/master/BulletData.json')
```

Comment:
Can we include this data in the repo itself? That way these tests don't break if that dataset is removed or if `rawgit` goes down.